### ISSUE TITLE
Graphs

### ISSUE THREAD

**Issue by tbreloff** [tbreloff/Plots.jl#48]
Just thinking out loud... when there's better support for line segments (#19) then I'll have the basics necessary for visualizing graphs.  I'd like to settle on the best representation of graphs ([LightGraphs](https://github.com/JuliaGraphs/LightGraphs.jl)?), and add the necessary hooks to add them to arbitrary plots, similar to how I'm doing DataFrames currently.  Defaults could include no ticks/grid, etc.

There are numerous benefits to hooking into the Plots pipeline early in graphs visualization... one of them is that we could extend any graphs code to 3D once that framework is in place, with no extra work.

**Comment by tbreloff** [comment]
Also as an extension, I would like to see easy visualizations of neural networks resulting from this effort, and complex visualizations of network structure, progress, statistics, etc.

**Comment by waldyrious** [comment]
For reference, the current Julia graph visualization packages (that I know of) are:
- [GraphViz](https://github.com/Keno/GraphViz.jl)
- [GraphLayout](https://github.com/IainNZ/GraphLayout.jl)
- [TikzGraphs](https://github.com/sisl/TikzGraphs.jl)
- [GraphPlot](https://github.com/afternone/GraphPlot.jl)

**Comment by tbreloff** [comment]
See also: https://groups.google.com/forum/?hl=en#!topic/julia-users/V_IWw8EaF4s

**Comment by Evizero** [comment]
:+1:

**Comment by tbreloff** [comment]
Here's how I left that julia-users thread:

> I'm willing to step up!  I just don't want to step on Iain's toes.  My plan would be:

> - strip out all drawing-related code
- simplify the interface
- add some "_apply_recipe" methods in Plots for Graphs.jl objects, which would abstract the drawing from both the layout and the backend

> Then the "lines, circles, boxes, text..." would live in Plots, and the low-level drawing commands are not needed.

@IainNZ I think the most straightforward path forward is for me to cherry pick the core algorithms from GraphLayout.jl and put it in a file `src/graphlayout.jl` which has a "requires block" around the include statement.  This would mean copy/pasting (with a comment for attribution) from GraphLayout into Plots.  @IainNZ are you ok with that?  I hesitate to move forward with the plan without your approval.  Thanks.

**Comment by IainNZ** [comment]
Definitely can't stop you copying those files out as long as you update your LICENSE.md and comment it, but, seems pretty strange to me. I assume you aren't going to include the tree algorithm, right? You just mean the force-based ones? But if so, don't the force-based-ones already separate plotting and layout?

**Comment by tbreloff** [comment]
@IainNZ I'm actually more interested in the tree layout right now.  I definitely don't need to copy anything... but you made the comment that you wanted to deprecate the package, so I figured that was one alternative.  Would you rather transfer the package somewhere?  Or would you actively accept PRs to re-design the current package? Advice welcome!

**Comment by IainNZ** [comment]
The tree layout isn't really done, so I wouldn't be looking at that first. Additionally, from the reading I used to write that code, considering how the tree will appear is fairly essential to the layout process (e.g. node size) so I'm not sure it can be separated.

I was thinking about deprecating, but probably won't. I don't want to transfer it. But would accept PRs that are consistent with my intentions for the package.

I guess I also don't really get what you are trying to achieve here - is Plots.jl meant to be some omni-package to plot every kind of data structure? I've never really seen graph layout and plotting tabular data in one package before. Is the main point that its decouple from a backend? But then Compose.jl is also decoupled - I can output to SVG, PDF, PNG...

**Comment by tbreloff** [comment]
> is Plots.jl meant to be some omni-package to plot every kind of data structure?

Sort of... when they are generic enough.  And if they aren't generic enough, then they can still belong in a sister package (like MLPlots.jl)  From my perspective, graph visualization should be much more standard/core to visualization libraries.  I want to be able to create animations of graph traversal algorithms, annotate sections, interact and explore, etc.  Sure, you can accomplish these things in Compose (I mean, that's all Gadfly does is map plots to Compose).  But what if you want to incorporate real-time physics in GLVisualize, or underlay a 3D surface representing terrain?  You'll probably have trouble doing that with Compose.  Plots should abstract the "common visualization tasks" so you're not locked into a specific drawing platform.  There's also a lot of similarities in defining graph viz attributes and plots... colors, shapes, annotations, etc.  Why not share the infrastructure?

> I've never really seen graph layout and plotting tabular data in one package before.

Yeah that seems silly to me.  I think the interface and graph layout are sticking points for library developers, and they just don't feel pressure to implement the features.  With Plots and GraphLayout, these things are mostly done already, and so you can plug in to ready-made solutions.

> But would accept PRs that are consistent with my intentions for the package.

Perfect... this is what I'll do then.  My thinking is just to separate the functionality so that instead of 1 big function which does a complete layout and draws it, there are several smaller functions to compile intermediate values (coordinates, etc) that I can call directly.  There is the spring layout that returns coordinates already, and I'd like to extend that concept.  In the end, I would hope/expect that the interface and functionality is not changed... just the internals.  Then I can build off those internals for generic visualization design.

**Comment by StefanKarpinski** [comment]
This seems to be where this is heading, but I just wanted to chime in to say that the right course for this, imo, would be to rehabilitate GraphLayout to the point where it can do what Plots needs rather than fracturing the ecosystem further by having lots of bits of code to do various aspects of graph layout in lots of different places.

**Comment by tbreloff** [comment]
Thanks @StefanKarpinski I agree completely. That is my preference too.

**Comment by tbreloff** [comment]
I need to spend some time on graph layout algos, but I have some support for graphs:

![tmp](https://cloud.githubusercontent.com/assets/933338/13612406/4da7e604-e535-11e5-8725-b266299fbd10.png)

It still requires too much manual intervention for my taste.

**Comment by tbreloff** [comment]
I started some graph work in https://github.com/JuliaPlots/PlotRecipes.jl/blob/master/src/graphs.jl... also see: https://github.com/JuliaPlots/RecipesBase.jl/issues/3.

@IainNZ If the graph recipes turn out well, I'd be interested in merging them into GraphLayout... replacing the Compose dependency with a dependency on RecipesBase.  (you could still draw to Compose... it would just pass through the Plots pipeline first)  I'll keep you posted on progress, and you can let me know how you'd like to proceed.  I'm perfectly happy to keep things separate from GraphLayout, but it would also be great to merge efforts there.

**Comment by tbreloff** [comment]
Closing in favor of "Recipe ideas" issue
